src/messaging/messaging-service.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * @module MessagingService
+ */
+import ResponseMessage from './messages/response-message'
+import StoredRequest from './stored-request'
+
+/** A messaging for sending and receiving messages to and from various destinations */
+export default class MessagingService {
+  constructor (destinations = []) {
+    /**
+     * A map object where outgoing messages will be stored. The key is the message ID and the value is an object
+     * that stores details about the message being sent.
+     *
+     * @type {Map<string, StoredRequest>}
+     */
+    this._messages = new Map()
+
+    /**
+     * A map object where outgoing messages will be stored. The key is a destination name and the value is
+     * the Destination object.
+     *
+     * @type {Map<string, Destination>}
+     */
+    this._destinations = new Map()
+
+    // If provided as a singular value convert destination into an array
+    if (!Array.isArray(destinations)) { destinations = [destinations] }
+    destinations.forEach(destination => this.registerDestination(destination))
+  }
+
+  /**
+   * Registers destination by adding it to the destinations map and setting a response callback.
+   *
+   * @param {Destination} destination - A destination object to register.
+   */
+  registerDestination (destination) {
+    this._destinations.set(destination.name, destination)
```

Comment:
should we have some error checking here to be sure we aren't overwriting an existing destination by the same name?

Reply:
I thought of having this as a hot replacement feature when we'll be able to replace destination "on the fly" (i.e. if it moved somewhere, for example, and we want to provide a new destination object with an updated location). It might also be helpful for testing: if we have a remote serer destination it might be beneficial to replace it with a local emulator during a test run by just hot-plugging an emulator in. What do you think about that idea?

Reply:
hmm. That's a good point. But maybe we should have an method or argument for that? i.e. updateDestination rather than registerDestination?  Or registerDestination(destination, replace=false). Just to be sure when it's overridden we mean to?

Reply:
I think two different methods for adding and updating is a good idea. I'll add a separate update method.

Reply:
Done that